drafts/bitcartel-nu0-network-handshake/draft.rst

```diff
@@ -0,0 +1,191 @@
+
+To assist network coalescence before the activation height, we update the eviction process to place a higher priority on evicting legacy nodes.
+
+This can be activated at n blocks before the activation block height, where n could be defined by a constant such as::
+
+    static const int NETWORK_COALESCE_BLOCK_PERIOD = 1000.
+
+The eviction code can be updated as follows::
+
+    static bool AttemptToEvictConnection(bool fPreferNewConnection) {
+    ...
+    // Protect connections with certain characteristics
+    ...
+    // Check version of eviction candidates
+    if (current_block_height >= (activationheight - NETWORK_COALESCE_BLOCK_PERIOD)) {
+      // if there exist any legacy nodes, keep them in the eviction set
+      // and at the same time remove overwinter nodes from eviction set.
+      // if there do not exist any legacy nodes,
+      // continue with existing behaviour.
+
+
+Disconnecting Existing Connections
+----------------------------------
+
+It is likely that at the activation block height, an Overwinter node will still be connected to some Legacy nodes.
+
+Currently, when connecting, a node must perform the networking handshake, and send the version message, before any other messages are processed.
+
+To disconnect existing connections, we can modify ProcessMessage so that the protocol version is always checked after Overwinter activates.
+
+Example code::
+
+    bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv, int64_t nTimeReceived)
+        ...
+        else if (pfrom->nVersion == 0)
+        {
+            // Must have a version message before anything else
+            Misbehaving(pfrom->GetId(), 1);
+            return false;
+        }
+        else if (strCommand == "verack")
+        {
+            ...
+        }
+
+        // Disconnect existing connection if:
+        // 1. The version message has been received
+        // 2. Overwinter is active
+        // 3. Version is legacy
+        else if (
+            pfrom->nVersion != 0 &&
+            isOverwinterActivated() &&
+            pfrom->nVersion < OVERWINTER_PROTO_VERSION )
+        {
+            LogPrintf("peer=%d using obsolete version %i; disconnecting\n", pfrom->id, pfrom->nVersion);
+            pfrom->PushMessage("reject", strCommand, REJECT_OBSOLETE,
+                               strprintf("Version must be %d or greater", OVERWINTER_PROTO_VERSION));
+            pfrom->fDisconnect = true;
+            return false;
+        }
+
+
+
+Deployment
+==========
+
+This proposal will be deployed with the Overwinter network upgrade.
+
+Testnet:
+
+Mainnet:
+
+Backward compatibility
+======================
+
+This proposal intentionally creates what is known as a "bilateral hard fork" between Legacy software and Overwinter compatible software. Use of this new handshake requires that all network participants upgrade their software to a compatible version within the upgrade window
```

Comment:
bilateral sounds like both sides reject the other, but you say legacy nodes will not reject overwinter nodes in this handshake.

Reply:
Good point, will clean up.